stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1375 @@
+            address: miner_addr,
+            recipient,
+            block_hash: block_hash.clone(),
+            consensus_hash: block_consensus_hash.clone(),
+            parent_block_hash: parent_block_hash.clone(),
+            parent_consensus_hash: parent_consensus_hash.clone(),
+            coinbase: coinbase_reward_ustx,
+            tx_fees: MinerPaymentTxFees::Nakamoto { parent_fees },
+            burnchain_commit_burn,
+            burnchain_sortition_burn,
+            miner: true,
+            stacks_block_height: block_height,
+            vtxindex: 0,
+        };
+
+        Ok(miner_reward)
+    }
+
+    /// Return the total ExecutionCost consumed during the tenure up to and including
+    ///  `block`
+    pub fn get_total_tenure_cost_at(
+        conn: &Connection,
+        block: &StacksBlockId,
+    ) -> Result<Option<ExecutionCost>, ChainstateError> {
+        let qry = "SELECT total_tenure_cost FROM nakamoto_block_headers WHERE index_block_hash = ?";
+        conn.query_row(qry, &[block], |row| row.get(0))
+            .optional()
+            .map_err(|e| ChainstateError::DBError(e.into()))
+    }
+
+    /// Return the total transactions fees during the tenure up to and including
+    ///  `block`
+    pub fn get_total_tenure_tx_fees_at(
+        conn: &Connection,
+        block: &StacksBlockId,
+    ) -> Result<Option<u128>, ChainstateError> {
+        let qry = "SELECT tenure_tx_fees FROM nakamoto_block_headers WHERE index_block_hash = ?";
+        let tx_fees_str: Option<String> =
+            conn.query_row(qry, &[block], |row| row.get(0)).optional()?;
+        tx_fees_str
+            .map(|x| x.parse())
+            .transpose()
+            .map_err(|_| ChainstateError::DBError(DBError::ParseError))
+    }
+
+    /// Return a Nakamoto StacksHeaderInfo at a given tenure height in the fork identified by `tip_index_hash`
+    pub fn get_header_by_tenure_height(
+        tx: &mut StacksDBTx,
+        tip_index_hash: &StacksBlockId,
+        tenure_height: u64,
+    ) -> Result<Option<StacksHeaderInfo>, ChainstateError> {
+        // query for block header info at the tenure-height, then check if in fork
+        let qry =
+            "SELECT * FROM nakamoto_block_headers WHERE tenure_changed = 1 AND tenure_height = ?";
+        let candidate_headers: Vec<StacksHeaderInfo> =
+            query_rows(tx.tx(), qry, &[u64_to_sql(tenure_height)?])?;
+
+        if candidate_headers.len() == 0 {
+            // no nakamoto_block_headers at that tenure height, check if there's a stack block header where
+            //   block_height = tenure_height
+            let ancestor_at_height = match tx
+                .get_ancestor_block_hash(tenure_height, tip_index_hash)?
+                .map(|ancestor| Self::get_block_header(tx.tx(), &ancestor))
+                .transpose()?
+                .flatten()
+            {
+                Some(x) => x,
+                None => return Ok(None),
+            };
```

Comment:
I did not know about this construct!

Reply:
The `let-else` construct was introduced almost a year ago in Rust 1.65. It can be used almost anywhere you have a `match` with `Some(x) => x` or `Ok(x) => x`. Unless we are trying to maintain compatibility with old versions of `rustc` (I don't know why we would...) we should prefer this for readability and smaller code